src/facet/selection/_parameters.py

```diff
@@ -0,0 +1,412 @@
+
+    def to_expression(self) -> Expression:
+        """[see superclass]"""
+        return self._to_expression([])
+
+    def _to_expression(self, path_prefix: Union[str, List[str]]) -> Expression:
+        # path_prefix: the path prefix to prepend to each parameter name
+
+        def _values_to_expression(values: ParameterSet) -> Expression:
+            if isinstance(values, rv_frozen):
+                values: rv_frozen
+                return Id(values.dist.name)(*values.args, **values.kwds)
+            elif isinstance(values, (stats.rv_continuous, stats.rv_discrete)):
+                try:
+                    return Id(values.name)(values.a, values.b)
+                except AttributeError:
+                    pass
+
+            return make_expression(values)
+
+        path_prefix = (
+            []
+            if path_prefix is None
+            else to_list(path_prefix, element_type=str, arg_name="path_prefix")
+        )
+
+        parameters = {
+            ".".join(path): _values_to_expression(value)
+            for path, value in self._iter_parameters(path_prefix=path_prefix)
+        }
+
+        if path_prefix:
+            return Id(type(self))(
+                **{".".join(path_prefix): self.estimator}, **parameters
+            )
+        else:
+            return Id(type(self))(self.estimator, **parameters)
+
+
+@inheritdoc(match="""[see superclass]""")
+class MultiEstimatorParameterSpace(
+    BaseParameterSpace[T_Estimator], Generic[T_Estimator]
+):
+    """
+    A collection of parameter spaces, each representing a competing estimator from which
+    select the best-performing candidate with optimal hyper-parameters.
+
+    See :class:`.ParameterSpace` for documentation on how to set up and use parameter
+    spaces.
+    """
+
+    STEP_CANDIDATE = "candidate"
+
+    #: The estimator base type which all candidate estimators must implement.
+    estimator_type: Type[T_Estimator]
+
+    def __init__(
+        self,
+        *candidates: ParameterSpace[T_Estimator],
+        estimator_type: Type[T_Estimator],
+    ) -> None:
+        """
+        :param candidates: the parameter spaces from which to select the best estimator
+        :param estimator_type: the estimator base type which all candidate estimators
+            must implement
+        """
+        validate_element_types(candidates, expected_type=ParameterSpace)
+        validate_candidates(candidates, expected_estimator_type=estimator_type)
+
+        if len(candidates) == 0:
+            raise TypeError("no parameter space passed; need to pass at least one")
+
+        if all(
+            isinstance(candidate.estimator, EstimatorDF) for candidate in candidates
+        ):
+            cls_pipeline = PipelineDF
+        else:
+            cls_pipeline = Pipeline
+
+        super().__init__(
+            estimator=cls_pipeline(
+                [(MultiEstimatorParameterSpace.STEP_CANDIDATE, candidates[0].estimator)]
```

Comment:
Hi @j-ittner, I've got first question: There's a thing that I don't fully understand and I wanted to be sure about it. When using `MultiEstimatorParameterSpace` and fitting `LearnerRanker2`, it sets up `estimator` as `candidates[0].estimator` but then on fitting it gets overridden by `"candidate"` argument in parameter space. Thanks to that we can evaluate multiple estimators on single `BaseSearchCV.fit` call without using multiple searchers. But then accessing `searcher.estimator` doesn't make sense because it's just one that gets overridden anyway.

My two thoughts:
1. Maybe let's make `searcher_` in `LearnerRanker2` protected (`_searcher`) because:
```python
ranker = LearnerRanker2(GridSearchCV, ps)
ranker.searcher_.estimator # <- makes sense as it's just `ps.estimator`

ranker = LearnerRanker2(GridSearchCV, multi_ps)
ranker.searcher_.estimator # <- doesn't really make sense as it's just estimator taken from the first candidate
```

2. Change `estimator` in `MultiEstimatorParameterSpace.__init__` to something that will indicate it's not meant to be used:
```python
estimator=cls_pipeline([(MultiEstimatorParameterSpace.STEP_CANDIDATE, None)]) # This will fail I think but maybe `DummyEstimator`?
```
Then if somebody would access `ranker.searcher_.estimator` sees that it isn't this type of ParameterSpace.

What do you think?